grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java

```diff
@@ -429,13 +415,39 @@ public EntityType putEntityType(Label label) {
                 v -> factory().buildEntityType(v, getMetaEntityType()));
     }
 
-    private <T extends SchemaConcept> T putSchemaConcept(Label label, Schema.BaseType baseType, Function<VertexElement, T> factory) {
+    /**
+     * This is a helper method which will either find or create a {@link SchemaConcept}.
+     * When a new {@link SchemaConcept} is created it is added for validation through it's own creation method for
+     * example {@link ai.grakn.kb.internal.concept.RoleImpl#create(VertexElement, Role, Boolean)}.
+     *
+     * When an existing {@link SchemaConcept} is found it is build via it's get method such as
+     * {@link ai.grakn.kb.internal.concept.RoleImpl#get(VertexElement)} and skips validation.
+     *
+     * Once the {@link SchemaConcept} is found or created a few checks for uniqueness and correct
+     * {@link ai.grakn.util.Schema.BaseType} are performed.
+     *
+     * @param label The {@link Label} of the {@link SchemaConcept} to find or create
+     * @param baseType The {@link Schema.BaseType} of the {@link SchemaConcept} to find or create
+     * @param newConceptFactory the factory to be using when creating a new {@link SchemaConcept}
+     * @param <T> The type of {@link SchemaConcept} to return
+     * @return a new or existing {@link SchemaConcept}
+     */
+    private <T extends SchemaConcept> T putSchemaConcept(Label label, Schema.BaseType baseType, Function<VertexElement, T> newConceptFactory) {
         checkSchemaMutationAllowed();
-        SchemaConcept schemaConcept = buildSchemaConcept(label, () -> factory.apply(putVertex(label, baseType)));
 
+        //Get the type if it already exists otherwise build a new one
+        SchemaConceptImpl schemaConcept = getSchemaConcept(convertToId(label));
+        if (schemaConcept == null) {
+            VertexElement vertexElement = addTypeVertex(getNextId(), label, baseType);
+            schemaConcept = SchemaConceptImpl.from(buildSchemaConcept(label, () -> newConceptFactory.apply(vertexElement)));
+        } else if (!baseType.equals(schemaConcept.baseType())) {
+            throw labelTaken(schemaConcept);
+        }
+
+        //Check if the type we got is correct
+        SchemaConceptImpl finalSchemaConcept = schemaConcept;
         T finalType = validateSchemaConcept(schemaConcept, baseType, () -> {
-            if (Schema.MetaSchema.isMetaLabel(label)) throw GraknTxOperationException.reservedLabel(label);
-            throw PropertyNotUniqueException.cannotCreateProperty(schemaConcept, Schema.VertexProperty.SCHEMA_LABEL, label);
+            throw labelTaken(finalSchemaConcept);
```

Comment:
Another optimisation: maybe we don't need to check this if the schema concept was just created.

Reply:
Maybe, but the label is cached so we wont get much from this. Will check though.